compiler/noirc_evaluator/src/ssa/opt/remove_unreachable_instructions.rs

```diff
@@ -0,0 +1,461 @@
+    ir::{
+        basic_block::BasicBlockId,
+        dfg::DataFlowGraph,
+        dom::DominatorTree,
+        function::Function,
+        instruction::{Instruction, TerminatorInstruction},
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    pub(crate) fn remove_unreachable_instructions(mut self) -> Ssa {
+        for function in self.functions.values_mut() {
+            function.remove_unreachable_instructions();
+        }
+        self
+    }
+}
+
+impl Function {
+    fn remove_unreachable_instructions(&mut self) {
+        let mut dom = DominatorTree::with_function(self);
+
+        // The current block we are currently processing
+        let mut current_block_id = None;
+
+        // Whether the current block instructions were determined to be unreachable
+        let mut current_block_instructions_are_unreachable = false;
+
+        // Blocks that can't be reached because their dominator has an always failing instruction.
+        let mut unreachable_blocks = HashSet::default();
+
+        self.simple_reachable_blocks_optimization(|context| {
+            let block_id = context.block_id;
+
+            if current_block_id != Some(block_id) {
+                current_block_id = Some(block_id);
+                current_block_instructions_are_unreachable = unreachable_blocks.contains(&block_id);
+            }
+
+            if current_block_instructions_are_unreachable {
+                context.remove_current_instruction();
+                return;
+            }
+
+            let instruction = context.instruction();
+            let always_failing_instruction = match instruction {
+                Instruction::Constrain(lhs, rhs, _) => {
+                    let Some(lhs_constant) = context.dfg.get_numeric_constant(*lhs) else {
+                        return;
+                    };
+                    let Some(rhs_constant) = context.dfg.get_numeric_constant(*rhs) else {
+                        return;
+                    };
+                    lhs_constant != rhs_constant
+                }
+                Instruction::ConstrainNotEqual(lhs, rhs, _) => {
+                    let Some(lhs_constant) = context.dfg.get_numeric_constant(*lhs) else {
+                        return;
+                    };
+                    let Some(rhs_constant) = context.dfg.get_numeric_constant(*rhs) else {
+                        return;
+                    };
+                    lhs_constant == rhs_constant
+                }
+                _ => false,
+            };
+
+            if always_failing_instruction {
+                current_block_instructions_are_unreachable = true;
+
+                add_dominated_blocks(block_id, context.dfg, &mut dom, &mut unreachable_blocks);
+                let terminator = context.dfg[block_id].take_terminator();
+                let call_stack = terminator.call_stack();
+                context.dfg[block_id]
+                    .set_terminator(TerminatorInstruction::Unreachable { call_stack });
```

Comment:
You will do this for each instruction in the block. Maybe it can be done only if there was a switch?

Reply:
I just checked and no, because we set `current_block_instructions_are_unreachable = true`, then this code above will be hit first:

```rust
            if current_block_instructions_are_unreachable {
                context.remove_current_instruction();
                return;
            }
```